EmbeddedSurfacesParallelSynchronization.cpp

```diff
@@ -0,0 +1,459 @@
+  packedSize += elemManager.PackGlobalMaps( sendBufferPtr, newElemsToSend );
+
+  packedSize += nodeManager.packUpDownMaps( sendBufferPtr, newNodesToSend );
+  packedSize += elemManager.PackUpDownMaps( sendBufferPtr, newElemsToSend );
+
+  packedSize += nodeManager.pack( sendBufferPtr, {}, newNodesToSend, 0, false, packEvents );
+  packedSize += elemManager.Pack( sendBufferPtr, {}, newElemsToSend );
+
+
+  GEOSX_ERROR_IF( bufferSize != packedSize, "Allocated Buffer Size is not equal to packed buffer size" );
+
+
+  GEOSX_UNUSED_VAR( commID );
+  neighbor->mpiISendReceive( commID, MPI_COMM_GEOSX );
+}
+
+void EmebeddedSurfacesParallelSynchronization::unpackNewToGhosts( NeighborCommunicator * const neighbor,
+                                                                  int commID,
+                                                                  MeshLevel & mesh )
+{
+  int unpackedSize = 0;
+
+  EmbeddedSurfaceNodeManager & nodeManager = mesh.getEmbSurfNodeManager();
+  ElementRegionManager & elemManager = mesh.getElemManager();
+
+  buffer_type const & receiveBuffer = neighbor->receiveBuffer( commID );
+  buffer_unit_type const * receiveBufferPtr = receiveBuffer.data();
+
+  localIndex_array newGhostNodes;
+  ElementRegionManager::ElementReferenceAccessor< localIndex_array > newGhostElems;
+  array1d< array1d< localIndex_array > > newGhostElemsData;
+  newGhostElems.resize( elemManager.numRegions() );
+  newGhostElemsData.resize( elemManager.numRegions() );
+
+
+  for( localIndex er=0; er<elemManager.numRegions(); ++er )
+  {
+    ElementRegionBase & elemRegion = elemManager.getRegion( er );
+    newGhostElemsData[er].resize( elemRegion.numSubRegions() );
+    newGhostElems[er].resize( elemRegion.numSubRegions() );
+    for( localIndex esr=0; esr<elemRegion.numSubRegions(); ++esr )
+    {
+      newGhostElems[er][esr].set( newGhostElemsData[er][esr] );
+    }
+  }
+
+  parallelDeviceEvents events;
+
+  unpackedSize += nodeManager.unpackGlobalMaps( receiveBufferPtr, newGhostNodes, 0 );
+  unpackedSize += elemManager.UnpackGlobalMaps( receiveBufferPtr, newGhostElems );
+
+  unpackedSize += nodeManager.unpackUpDownMaps( receiveBufferPtr, newGhostNodes, true, true );
+  unpackedSize += elemManager.UnpackUpDownMaps( receiveBufferPtr, newGhostElems, true );
+
+  unpackedSize += nodeManager.unpack( receiveBufferPtr, newGhostNodes, 0, false, events );
+  unpackedSize += elemManager.Unpack( receiveBufferPtr, newGhostElems );
+
+  waitAllDeviceEvents( events );
+
+  elemManager.forElementSubRegionsComplete< ElementSubRegionBase >(
+    [&]( localIndex const er, localIndex const esr, ElementRegionBase &, ElementSubRegionBase & subRegion )
+  {
+    localIndex_array & elemGhostsToReceive = subRegion.getNeighborData( neighbor->neighborRank() ).ghostsToReceive();
+
+    if( newGhostElemsData[er][esr].size() > 0 )
+    {
+      elemGhostsToReceive.move( LvArray::MemorySpace::CPU );
```

Comment:
Yeah. Honestly this should be put anywhere you touch a view that hasn't been captured in a RAJA loop.

When an `ArrayView` gets moved to device it's pointer gets set to a device pointer, but the parent `Array` still has the host pointer. So for something like `elemGhostsToReceive` which is never modified on device it will work OK to just access the data without capturing or moving if you don't modify it. But if you do modify it or if somewhere else someone calls `Array::move( GPU )` (in which case the `Array`'s pointer does get set to the device pointer) then you're screwed.